external-crates/move/move-vm/profiler/src/lib.rs

```diff
@@ -0,0 +1,307 @@
+pub struct FrameName {
+    name: String,
+}
+
+#[cfg(debug_assertions)]
+#[derive(Debug, Clone, Serialize)]
+pub struct Shared {
+    frames: Vec<FrameName>,
+
+    #[serde(skip)]
+    frame_table: BTreeMap<String, usize>,
+}
+
+#[cfg(debug_assertions)]
+#[derive(Debug, Clone, Serialize)]
+pub struct Event {
+    #[serde(rename(serialize = "type"))]
+    ty: String,
+    frame: u64,
+    at: u64,
+}
+
+#[cfg(debug_assertions)]
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Profile {
+    #[serde(rename(serialize = "type"))]
+    ty: String,
+    name: String,
+    unit: String,
+    start_value: u64,
+    end_value: u64,
+    events: Vec<Event>,
+}
+
+#[cfg(debug_assertions)]
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct GasProfiler {
+    exporter: String,
+    name: String,
+    active_profile_index: u64,
+    #[serde(rename(serialize = "$schema"))]
+    schema: String,
+    shared: Shared,
+    profiles: Vec<Profile>,
+
+    #[serde(skip)]
+    pub start_gas: u64,
+    #[serde(skip)]
+    pub config: VMProfilerConfig,
+}
+
+#[cfg(debug_assertions)]
+impl GasProfiler {
+    pub fn init(config: &VMProfilerConfig, name: String, start_gas: u64) -> Self {
+        let mut prof = GasProfiler {
+            exporter: "speedscope@1.15.2".to_string(),
+            name: name.clone(),
+            active_profile_index: 0,
+            schema: "https://www.speedscope.app/file-format-schema.json".to_string(),
+            shared: Shared {
+                frames: vec![],
+                frame_table: BTreeMap::new(),
+            },
+            profiles: vec![Profile {
+                ty: "evented".to_string(),
+                name,
+                unit: "none".to_string(),
+                start_value: 0,
+                end_value: 0,
+                events: vec![],
+            }],
+            start_gas,
+            config: config.clone(),
+        };
+        profile_open_frame_impl!(Some(&mut prof), "root".to_string(), start_gas);
+        prof
+    }
+
+    pub fn init_default_cfg(name: String, start_gas: u64) -> Self {
+        Self::init(&VMProfilerConfig::default(), name, start_gas)
+    }
+
+    fn profile_name(&self) -> String {
+        self.name.clone()
+    }
+
+    pub fn short_name(s: &String) -> String {
+        s.split("::").last().unwrap_or(s).to_string()
+    }
+
+    fn is_metered(&self) -> bool {
+        (self.profiles[0].end_value != 0) && (self.start_gas != 0)
+    }
+
+    fn start_gas(&self) -> u64 {
+        self.start_gas
+    }
+
+    fn add_frame(&mut self, frame_name: String) -> u64 {
+        *self
+            .shared
+            .frame_table
+            .entry(frame_name.clone())
+            .or_insert({
+                let val = self.shared.frames.len() as u64;
+                self.shared.frames.push(FrameName { name: frame_name });
+                val as usize
+            }) as u64
+    }
+
+    #[cfg(debug_assertions)]
```

Comment:
Nits: do these functions need the `debug_assertions` macro on them since they're inside an impl block that is also macro-gated?

Reply:
Fixed